### PR TITLE
Fix: add recursion and memory limits to parser to prevent DoS attacks CLAUDE

### PR DESCRIPTION
**Pull request by ori-n**
## Summary

Adds `max_depth` and `max_nodes` parameters to the Parser to prevent Denial of Service attacks from pathological queries that could cause stack overflow or memory exhaustion.

- **max_depth** (default: 120): Prevents deeply nested queries with 50+ levels of subqueries from causing RecursionError
- **max_nodes** (default: 1M): Limits total AST nodes to prevent memory exhaustion from engineered malicious queries

## Implementation

- Added depth tracking with `_parse_depth()` context manager
- Added node counting to enforce memory limits
- Both limits raise ParseError when exceeded, handled by existing error handling
- Comprehensive tests in `test_parser_dos_protection.py`

## Test Coverage

Added 960+ lines of tests covering:
- Stack overflow prevention with deeply nested queries
- Memory exhaustion prevention with pathological node generation
- Edge cases and boundary conditions

🤖 Generated with Claude Code

### CODE REVIEW

**Comment by tobymao** [comment]
this should be off by default. if not see we should never hit these paths so that there are no performance implications

**Comment by ori-n** [comment]
The defaults are very permissive: 
- max_depth: 120 (typical queries rarely exceed 10-15 levels)
- max_nodes: 1,000,000 (typical queries have <100k nodes)

Performance impact is minimal since depth/node tracking only involves simple integer operations on normal paths.
Both are configurable:                                    
`
parse_one(sql, max_depth=200, max_nodes=5000000)
`                                                                                                                                       
I believe applications should set much lower limits based on their needs. This lets each application enforce its own security boundary while maintaining backward compatibility.

File: sqlglot/parser.py [hunk 1]
```diff
@@ -2239,28 +2296,33 @@ def _parse_whileblock(self) -> exp.WhileBlock:
         )
 
     def _parse_statement(self) -> exp.Expr | None:
-        if not self._curr:
-            return None
+        with self._parse_depth():
```
Comment (georgesittas) [inline sqlglot/parser.py:2299]: @ori-n what is the definition of "depth" in this case? As in, what kind of depth are you counting? Right now, there are a few calls of `_parse_depth()` that appear to be placed arbitrarily, leading to an unclear notion of depth, imo.

I think we should be more explicit about what cases we're guarding against exactly. You can end up with deep trees from many different inputs. For example, long chains of binary operators, function calls, subqueries, etc.

Also, different code paths put different pressure on the call stack because of the differences in call chains.

I'm wondering if we should just check for "subquery" depth or something instead of trying to make an unhelpful generalization. Or, we could just keep node count. I'm not sure which is preferrable here, yet.

cc @tobymao

Comment (ori-n) [reply sqlglot/parser.py:2299]: You're right, the depth definition is unclear as-is. The intent was to have two separate limits for two different attack vectors: max_nodes for AST             
  amplification/memory exhaustion, and max_depth for stack overflow from deeply nested input. Right now the depth check only covers some of the code paths, so things   
  like deeply nested CASE chains bypass it. I can work to cover more scenarios and make it more consistent, but the general idea stays the same. WDYT?

Comment (georgesittas) [reply sqlglot/parser.py:2299]: Thanks for clarifying. I'm not sure about this part:

> stack overflow from deeply nested input.

Python's `RecursionError` already catches stack overflow, and it's always active. The `max_nodes` flag likely has some value add as you say, though. I don't see the reason at the moment to have messy depth checks all over our parser code.

**Comment by georgesittas** [comment]
After discussing, we concluded that we're only interested in the node count threshold for the time being. I'm preparing a PR that is a cleaned up version of this, so I'm closing this one. In any case, thank you for the contribution.

**Comment by ori-n** [comment]
The Python recursion limit is an implementation detail that varies between environments and Python versions - it's not a reliable or user-facing contract. The depth limit is meant to be a logical SQL-level property that users can configure (e.g. "reject queries nested more than N levels deep"), with a clean ParseError rather than an unpredictable RecursionError. 
I'm glad you see value in max_nodes - I designed the two limits since they address different problems, and my suggestion is to include both. If you have a better approach for the depth limiting, I'll be happy to know about it.
